internal/provider/data_source_certificate.go

```diff
@@ -2,169 +2,221 @@ package provider
 )
 
-func dataSourceCertificate() *schema.Resource {
-	return &schema.Resource{
-		ReadContext: dataSourceCertificateRead,
+type (
+	certificateDataSourceType struct{}
+	certificateDataSource     struct {
+		provider *provider
+	}
+)
 
-		Description: "Get information about the TLS certificates securing a host.\n\n" +
-			"Use this data source to get information, such as SHA1 fingerprint or serial number, " +
-			"about the TLS certificates that protects a URL.",
+var (
+	_ tfsdk.DataSourceType = (*certificateDataSourceType)(nil)
+	_ tfsdk.DataSource     = (*certificateDataSource)(nil)
+)
 
-		Schema: map[string]*schema.Schema{
+func (dst *certificateDataSourceType) GetSchema(_ context.Context) (tfsdk.Schema, diag.Diagnostics) {
+	return tfsdk.Schema{
+		Attributes: map[string]tfsdk.Attribute{
+			// Required attributes
 			"url": {
-				Type:     schema.TypeString,
+				Type:     types.StringType,
 				Optional: true,
-				Description: "URL of the endpoint to get the certificates from. " +
-					fmt.Sprintf("Accepted schemes are: `%s`. ", strings.Join(SupportedURLSchemesStr(), "`, `")) +
+				Validators: []tfsdk.AttributeValidator{
+					attribute_validation.UrlWithScheme(supportedURLSchemesStr()...),
+					attribute_validation.ExactlyOneOf(
+						tftypes.NewAttributePath().WithAttributeName("content"),
+						tftypes.NewAttributePath().WithAttributeName("url"),
+					),
+				},
+				MarkdownDescription: "URL of the endpoint to get the certificates from. " +
+					fmt.Sprintf("Accepted schemes are: `%s`. ", strings.Join(supportedURLSchemesStr(), "`, `")) +
 					"For scheme `https://` it will use the HTTP protocol and apply the `proxy` configuration " +
 					"of the provider, if set. For scheme `tls://` it will instead use a secure TCP socket.",
-				ValidateDiagFunc: validation.ToDiagFunc(validation.IsURLWithScheme(SupportedURLSchemesStr())),
-				ExactlyOneOf:     []string{"content", "url"},
 			},
 			"content": {
-				Type:         schema.TypeString,
-				Optional:     true,
-				Description:  "The content of the certificate in [PEM (RFC 1421)](https://datatracker.ietf.org/doc/html/rfc1421) format.",
-				ExactlyOneOf: []string{"content", "url"},
+				Type:     types.StringType,
+				Optional: true,
+				Validators: []tfsdk.AttributeValidator{
+					attribute_validation.ExactlyOneOf(
+						tftypes.NewAttributePath().WithAttributeName("content"),
+						tftypes.NewAttributePath().WithAttributeName("url"),
+					),
+				},
+				MarkdownDescription: "The content of the certificate in [PEM (RFC 1421)](https://datatracker.ietf.org/doc/html/rfc1421) format.",
 			},
+
+			// Optional attributes
 			"verify_chain": {
-				Type:          schema.TypeBool,
-				Optional:      true,
-				Default:       true,
-				Description:   "Whether to verify the certificate chain while parsing it or not (default: `true`).",
-				ConflictsWith: []string{"content"},
+				Type:     types.BoolType,
+				Optional: true,
+				//Default:       true,
```

Comment:
If `verify_chain` should have a default value should it be computed and have a `PlanModifier` ?

Reply:
Mate, you actually spotted a bug!

You know what, I think it's time I add that test to fetch an actual live cert during a test, that @bflad suggested in the past: that would have spotted this issue immediately.

This would have been a guaranteed bugfix after release: thanks for intercepting this!